move-mutator/src/coverage.rs

```diff
@@ -0,0 +1,221 @@
+                    let first_location = func_source_coverage.uncovered_locations.first();
+
+                    // If not empty, get the uncovered spans for the function.
+                    let uncovered_spans = if let Some(loc) = first_location {
+                        // Can't fail since we already inserted this value earlier above.
+                        let file_contents = sources.get(&src_path).unwrap();
+
+                        let file_hash = loc.file_hash();
+                        UncoveredSpans::new(func_source_coverage, file_hash, file_contents)
+                    } else {
+                        continue;
+                    };
+
+                    let module = module.self_name();
+                    let function = identifer.into_string();
+                    let associated_fn_name = format!("{module}::{function}");
+                    all_uncovered_spans.insert(associated_fn_name, uncovered_spans);
+                }
+            });
+
+        trace!("all uncovered spans: {all_uncovered_spans:?}");
+        self.all_uncovered_spans = all_uncovered_spans;
+        Ok(())
+    }
+
+    /// Check if the location is covered by the unit test.
+    pub(crate) fn check_location(&self, associated_fn_name: String, loc: &Loc) -> bool {
+        let span = loc.span();
+
+        let Some(UncoveredSpans(spans)) = self.all_uncovered_spans.get(&associated_fn_name) else {
+            trace!("location has coverage since {associated_fn_name} has full coverage");
+            return true;
+        };
+
+        for uncovered_span in spans {
+            // Skip all early spans.
+            if span.start() >= uncovered_span.end() {
+                continue;
+            }
+
+            // Check if the span starts earlier than the uncovered span.
+            if uncovered_span.start() > span.start() {
+                break;
+            }
+
+            // If the span goes beyond the uncovered span, we are done here.
+            if uncovered_span.end() < span.end() {
+                break;
+            }
+
+            trace!("{associated_fn_name} has no coverage for the given location");
+            return false;
+        }
+
+        trace!("{associated_fn_name} has coverage for the given location");
+        true
+    }
+}
+
+#[derive(Debug)]
+struct UncoveredSpans(Vec<Span>);
+
+impl UncoveredSpans {
+    /// Create a new [`UncoveredSpans`].
+    fn new(func_src_cov: FunctionSourceCoverage, file_hash: FileHash, file_contents: &str) -> Self {
+        let merged_spans = merge_spans(file_hash, func_src_cov);
+        let optimized_spans = merge_spans_after_removing_whitespaces(merged_spans, file_contents);
+        Self(optimized_spans)
+    }
+}
+
+// This is based on the same function in aptos-core, but that one needs tweaking - it doesn't merge
+// spans properly in all cases, but that's not noticeable in the coverage output. We could port the
```

Comment:
What are those cases?

Reply:
This is the original logic:
```rust
    for interval in covs {
        if curr.disjoint(interval) {
            unioned.push(curr);
            curr = interval;
        } else {
            curr = curr.merge(interval);
        }
    }
```
This won't join spans like these:
`Span(0..2]` and `Span(2..5]`
since it will state that these are disjoint...
I can add information about this in the comment.

Done